src/index.js

```diff
@@ -29,369 +24,338 @@ class IPLDResolver {
-              }
-              return base
-            })))
-          })
-          return deferred
-        }),
-        pull.map((e) => {
-          if (typeof e === 'string') {
-            return e
+        // Only follow links if recursion is intended
+        if (options.recursive) {
+          cid = await maybeRecurse(block, treePath)
+          if (cid !== null) {
+            queue.push({ cid, basePath: fullPath + '/' })
           }
-          return e.basePath
-        }),
-        pull.filter(Boolean)
-      )
-    }
+        }
 
-    // filter out by path
-    if (path) {
-      return pull(
-        p,
-        pull.map((el) => {
-          if (el.indexOf(path) === 0) {
-            el = el.slice(path.length + 1)
-            return el
+        // Return it if it matches the given offset path, but is not the
+        // offset path itself
+        if (fullPath.startsWith(offsetPath) &&
+            fullPath.length > offsetPath.length) {
+          if (offsetPath.length > 0) {
+            fullPath = fullPath.slice(offsetPath.length + 1)
           }
-        }),
-        pull.filter(Boolean)
-      )
-    }
 
-    return p
-  }
+          yield fullPath
+        }
+      }
+    }.bind(this)
 
-  remove (cids, callback) {
-    this.bs.delete(cids, callback)
+    return extendIterator(generator())
   }
 
   /*           */
   /* internals */
   /*           */
+  async _getFormat (codec) {
+    // TODO vmx 2019-01-24: Once all CIDs support accessing the codec code
+    // instead of the name, remove this part
+    if (typeof codec === 'string') {
+      const constantName = codec.toUpperCase().replace(/-/g, '_')
+      codec = multicodec[constantName]
+    }
 
-  _get (cid, callback) {
-    waterfall([
-      (cb) => this._getFormat(cid.codec, cb),
-      (format, cb) => this.bs.get(cid, (err, block) => {
-        if (err) return cb(err)
-        cb(null, format, block)
-      }),
-      (format, block, cb) => {
-        format.util.deserialize(block.data, (err, deserialized) => {
-          if (err) {
-            return cb(err)
-          }
-          cb(null, deserialized)
-        })
-      }
-    ], callback)
-  }
-
-  _getFormat (codec, callback) {
     if (this.resolvers[codec]) {
-      return callback(null, this.resolvers[codec])
+      return this.resolvers[codec]
     }
 
     // If not supported, attempt to dynamically load this format
-    this.support.load(codec, (err, format) => {
-      if (err) return callback(err)
-      this.resolvers[codec] = format
-      callback(null, format)
-    })
+    const format = await this.loadFormat(codec)
+    this.addFormat(format)
+    return format
   }
 
-  _put (cid, node, callback) {
-    callback = callback || noop
+  async _store (cid, node) {
+    const format = await this._getFormat(cid.codec)
+    const serialized = await promisify(format.util.serialize)(node)
+    const block = new Block(serialized, cid)
+    await promisify(this.bs.put.bind(this.bs))(block)
+  }
```

Comment:
I think this is ok for now but the block store needs to take a iterator and return an iterator that yields the blocks you put in:

```js
put (nodes) {
  return this.bs.put(serialize(nodes))
}
```